_posts/2016-05-10-finite-state-machines-with-akka.md

```diff
@@ -0,0 +1,205 @@
+---
+layout: post
+name: Finite-state-machines-with-Akka
+title: 'Finite state machines with Akka'
+date: 2016-05-10 00:20:00 +00:00
+author: Felipe Fernández
+image:
+    src: /assets/img/custom/blog/machine.png
+    attribution:
+      text: Machine by Olle Svensson  
+      href: https://www.flickr.com/photos/ollesvensson/3335131082
+tags:
+- akka
+- state
+- fsm
+
+---
+
+Let's start with the architectural diagram contained in [the first post of the series.](http://codurance.com/2016/04/28/async-systems-with-sync-clients)
+
+<img src="/assets/img/custom/blog/law_enforcement.png" alt="Law enforcement architecture" title="Law enforcement architecture" class="img img-center img-responsive style-screengrab">
+
+We can understand the system through an example. The police sends us a request to delete an illegal item, and it expects a response in 10 seconds. Relevant status, for this example, are:
+
+* 200: the item has been successfully deleted in every container that it was published.
+* 404: the item doesn't exist in our system.
+* 504: timeout trying to delete the item.
+
+Law enforcement service communicates with Items service asynchronously using [Kafka](http://kafka.apache.org/). That means that we need to subscribe to a [topic](http://kafka.apache.org/documentation.html#intro_topics) called `item_deleted`. To add complexity to the system, we need to handle some multiplexing as the item could be published in different containers as the personal timeline or different groups. There is some internal state to keep inside of the lifecycle of a single request. Let's define what we mean with state, before of getting into the details of our solution.
+
+## Defining State
+
+State is the ability to keep track of what happened in our system. A pure stateless application would be a pure function that doesn't have any side effect. It receives an input, transform it following some rules and return an output. Those stateless applications are not very useful in a business context. Business and users want to know what happened in the past, so they can take informed decisions.
```

Comment:
...transforms it...returns an output.

Such stateless applications...

so that they can make informed decisions.